src/csharp/Grpc.Core.Api/BufferMarshaller.cs

```diff
@@ -0,0 +1,68 @@
+#region Copyright notice and license
+
+// Copyright 2015 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using Grpc.Core.Utils;
+
+namespace Grpc.Core
+{
+    /// <summary>
+    /// Encapsulates the logic for serializing and deserializing messages.
+    /// </summary>
+    public class BufferMarshaller<T> : Marshaller<T>
```

Comment:
Still considering alternative to creating a subclass.

Reply:
I don't what what value does it have to have a subclass like this, doesn't seem that the generated code gets any simpler?

Reply:
I agree. Getting the bufferwriter/sequence from the context is only a method call. A helper method would be better way to minimize generated code.

Reply:
Hmm, a helper method that itself isn't generated probably isn't possible. I thought of something like this:

```cs
public static void WriteMessage(SerializationContext ctx, IBufferMessage msg)
{
  // serialization
  ctx.SetPayloadLength(msg.CalculateSize());
  msg.writeTo(ctx.GetBufferWriter());
  ctx.Complete();
}
```

But it would need to know about gRPC's `(De)SerializationContext` and Protobuf's `IBufferMessage` types. The only place that knows about both assemblies is generated code.

Reply:
Right, I couldn't isolate a helper method by itself and needed something in the generated code.